lib/temporalio/worker/runner.rb

```diff
@@ -0,0 +1,58 @@
+module Temporalio
+  class Worker
+    class Runner
+      def initialize(*workers)
+        if workers.empty?
+          raise ArgumentError, 'Must be initialized with at least one worker'
+        end
+
+        @workers = workers
+        @mutex = Mutex.new
+        @started = false
+        @shutdown = false
+      end
+
+      def run(&block)
+        @thread = Thread.current
+        @started = true
+        workers.each { |worker| worker.start(self) }
+
+        block ? block.call : sleep
+      rescue Temporalio::Error::WorkerShutdown
+        # Explicit shutdown requested, no need to raise
+      ensure
+        @shutdown = true
+        shutdown_workers
+      end
+
+      def shutdown(exception = Temporalio::Error::WorkerShutdown.new('Manual shutdown'))
+        mutex.synchronize do
+          return unless running?
+
+          @shutdown = true
+        end
+
+        # propagate shutdown to the running thread
+        thread&.raise(exception)
+      end
+
+      private
+
+      attr_reader :workers, :mutex, :thread
+
+      def running?
+        @started && !@shutdown
+      end
+
+      def shutdown_workers
+        # Protect shutdown from any outside raises
+        Thread.handle_interrupt(StandardError => :never) do
+          workers.map do |worker|
+            # Shut down each worker (and wait for it) concurrently in separate threads
```

Comment:
Hrmm, anyway you can notify them to shutdown and then wait on them to complete? It may require reworking `Worker.shutdown` to not consume an entire thread.

Reply:
Yep, you can call `shutdown` on a `Worker::Runner`, but you'll need to figure out how to "wait" for it to complete (via a queue or some other means). I think it's fine to keep `Worker#shutdown` blocking. Alternatively you can also listen on a queue from the "until" block, so there's very little need to call shutdown explicitly

Reply:
It would definitely be nicer if we could shut them all down concurrently.

Reply:
Tracking this here — https://github.com/temporalio/sdk-ruby/issues/104